File: TinyScript/js/src/main/js/parser/ast/Expr.ts
import ASTNode from "./ASTNode";
import PeekTokenIterator from "../PeekTokenIterator";
import {PriorityTable} from "./PriorityTable";
import Token from "../../lexer/Token";
import ASTNodeTypes from "./ASTNodeTypes";
import TokenType from "../../lexer/TokenType";
import Factor from "./Factor";

export default class Expr extends ASTNode {

    constructor(parent: ASTNode | null) {
        super(null, null);

    }

    static TABLE = new PriorityTable();


    private static E(it: PeekTokenIterator, k: number): ASTNode | null {
        if (k < Expr.TABLE.table.length - 1) {
            return Expr.combine(it, () => this.E(it, k + 1), () => this.E_(it, k))
        } else {
            return Expr.race(it,
                () => Expr.combine(it, () => Expr.F(it), () => Expr.E_(it, k)),
                () => Expr.combine(it, () => Expr.U(it), () => Expr.E_(it, k))
            );
        }
    }

    private static race(it: PeekTokenIterator, aFunc: () => ASTNode | null, bFunc: () => ASTNode | null): ASTNode | null {
        if (!it.hasNext()) {
            return null;
        }
        const a = aFunc();
        if (a !== null) {
            return a;
        }
        return it.hasNext() ? bFunc() : null;
    }

    private static F(it: PeekTokenIterator): ASTNode | null {
        return Factor.parse(it);
    }

    private static U(it: PeekTokenIterator): ASTNode | null {
        const peek = it.peek();
        if (peek == null) {
            return null;
        }
        const value = peek.value;
        if (value == "(") {
            it.nextMatch("(");
            const e = Expr.E(it, 0);
            it.nextMatch(")");
            return e;
        } else if (value === "++" || value === "--" || value === "!") {
            it.next();
            const e1 = Expr.E(it, 0);
            const expr = Expr.fromToken(ASTNodeTypes.UNARY_EXPR, new Token(TokenType.OPERATOR, value));
            const e2 = Expr.E(it, 0);
            if (e2 != null) {
                expr.addChild(e2);
            }
            return null;
        }
        return null;
    }

    private static combine(it: PeekTokenIterator, aFunc: () => ASTNode | null, bFunc: () => ASTNode | null): ASTNode | null {
        if (!it.hasNext()) {
            return null;
        }
        const a = aFunc();
        if (a == null) {
            return it.hasNext() ? bFunc() : null;
        }
        const b = it.hasNext() ? bFunc() : null;
        if (b == null) {
            return a;
        }
        const expr = Expr.fromToken(ASTNodeTypes.BINARY_EXPR, new Token(TokenType.OPERATOR, b.label));
        expr.addChild(a);
        expr.addChild(b.children[0]);
        return expr;
    }

    private static E_(it: PeekTokenIterator, k: number): ASTNode | null {
        const token = it.peek();
        const value = token?.value;
        if (value == null) {
            return null;
        }
        if (this.TABLE.table[k].includes(value)) {
            it.next();
            const expr = Expr.fromToken(ASTNodeTypes.UNARY_EXPR, new Token(TokenType.OPERATOR, value))
            const combine = this.combine(it, () => this.E(it, k + 1), () => this.E_(it, k));
            if (combine != null) {
                expr.addChild(combine);
            }
            return expr;
        }
        return null;
    }

    public static parse(it: PeekTokenIterator): ASTNode | null {
        return Expr.E(it, 0);
    }

    private static fromToken(type: ASTNodeTypes, token: Token): Expr {
        const expr = new Expr(null);
        expr.label = token.value;
        expr.lexeme = token;
        expr.type = type;
        return expr;
    }
}

File: TinyScript/java/src/main/java/paser/ast/Expr.java
package paser.ast;

import lexer.Token;
import paser.util.ExprHOF;
import paser.util.ParseException;
import paser.util.PeekTokenIterator;
import paser.util.PriorityTable;

/**
 * @author MFine
 * @version 1.0
 * @date 2021/8/23 0:20
 **/
public class Expr extends ASTNode {


    private static PriorityTable table = new PriorityTable();


    protected Expr() {
        super();
    }

    public Expr( ASTNodeTypes type, Token lexeme) {
        super();
        this.type = type;
        this.lexeme = lexeme;
        this.label = lexeme.getValue();
    }


    // left: E(k) -> E(k) op(k) E(k+1) | E(k+1)
    // right:
    // E(k) -> E(K+1)E_(k)
    //    var e = new Expr();e.left = E(k+1);e.op = op(K);e.right=E(k+1)E_(k)
    // E_(k)->op(k)E(k+1)E_(k)|ε
    //最高优先级处理
    //  E(t)->F E_(k)|U E_(k)
    //  E_(t)-> op(t) E(t)E_(t) | ε
    //
    public static ASTNode E(ASTNode parent, int k, PeekTokenIterator it) throws ParseException {
        if (k < table.size() - 1) {
            return combine(parent, it, () -> E(parent, k + 1, it), () -> E_(parent, k, it));
        } else {
            return race(it,
                    () -> combine(parent, it, () -> U(parent, it), () -> E_(parent, k, it)),
                    () -> combine(parent, it, () -> F(parent, it), () -> E_(parent, k, it))
            );
        }
    }

    public static ASTNode U(ASTNode parent, PeekTokenIterator it) throws ParseException {
        Token token = it.peek();
        String value = token.getValue();
        ASTNode expr = null;

        if ("(".equals(value)) {
            it.nextMatch("(");
            expr = E(parent, 0, it);
            it.nextMatch(")");
            return expr;
        } else if ("++".equals(value) || "--".equals(value) || "!".equals(value)) {
            var t = it.peek();
            it.nextMatch(value);
            Expr unaryExpr = new Expr(ASTNodeTypes.UNARY_EXPR, t);
            unaryExpr.addChild(E(unaryExpr, 0, it));
            return unaryExpr;
        }
        return null;
    }

    public static ASTNode F(ASTNode parent, PeekTokenIterator it) throws ParseException {
        ASTNode parse = Factor.parse(parent, it);
        Token lookahead = it.peek();
        if (parse instanceof  Variable && "(".equals(lookahead.getValue())){
            ASTNode node = CallStmt.parse(parent, it);
            parse.addChild(node);
        }
        return parse;
    }

    private static ASTNode race(PeekTokenIterator it, ExprHOF aFunc, ExprHOF bFunc) throws ParseException {
        if (!it.hasNext()) {
            return null;
        }
        var a = aFunc.hoc();
        if (a != null) {
            return a;
        }
        return bFunc.hoc();
    }

    private static ASTNode E_(ASTNode parent, int k, PeekTokenIterator it) throws ParseException {
        var token = it.peek();
        var value = token.getValue();

        if (table.get(k).contains(value)) {
            Expr expr = new Expr(ASTNodeTypes.BINARY_EXPR, it.nextMatch(value));
            expr.addChild(combine(parent, it,
                    () -> E(parent, k + 1, it),
                    () -> E_(parent, k, it)
            ));
            return expr;
        }
        return null;
    }


    private static ASTNode combine(ASTNode parent, PeekTokenIterator it, ExprHOF aFunc, ExprHOF bFunc) throws ParseException {
        var a = aFunc.hoc();
        if (a == null) {
            return it.hasNext() ? bFunc.hoc() : null;
        }
        var b = it.hasNext() ? bFunc.hoc() : null;
        if (b == null) {
            return a;
        }
        Expr expr = new Expr(ASTNodeTypes.BINARY_EXPR, b.lexeme);
        expr.addChild(a);
        expr.addChild(b.getChildren(0));
        return expr;
    }

    public static ASTNode parse(PeekTokenIterator tokenIt) throws ParseException {
        return E(null, 0, tokenIt);
    }
}
